src/yetibot/core/adapters/discord.clj

```diff
@@ -0,0 +1,196 @@
+             user-model
+             :message
+             @yetibot-user
+             {:body message})))
+        (debug "Message from Yetibot => ignoring")))))
+
+
+
+(defn start
+  "start the discord connection"
+  [adapter _conn config _connected? bot-id yetibot-user]
+  (debug "starting discord connection")
+
+  (binding [*adapter* adapter]
+    (let [event-channel (chan 100)
+          message-channel (discord-ws/connect-bot! (:token config) event-channel :intents #{:guilds :guild-messages :guild-message-reactions :direct-messages :direct-message-reactions})
+          rest-connection (start-connection! (:token config))]
+      (let [retcon {:event  event-channel
+                    :message message-channel
+                    :rest    rest-connection}]
+        (reset! _conn retcon)
+
+        (debug (pr-str _conn))
+
+        (reset! _connected? true)
+        (reset! bot-id {:id @(get-current-user! rest-connection)})
+        (reset! yetibot-user @(get-current-user! rest-connection))
+        (message-pump! event-channel (fn [event-type event-data] (handle-event event-type event-data _conn yetibot-user)))))))
+
+
+
+(defn- channels [a]
+  (let [guild-channels (get-guild-channels!)]
+    (debug "Guild Channels: " (pr-str guild-channels))
+    (guild-channels)))
+
+(defn- send-msg [adapter msg conn]
+  (debug "Trying to send message: " msg)
+  (debug "Target is: " *target*)
+  (debug "Adapter: " (pr-str adapter))
+  (debug "conn: " (pr-str conn))
+  (debug "rest: " (pr-str (:rest @conn)))
+  (create-message! (:rest @conn) *target* :content msg))
+
+(defn stop
+  "stop the discord connection"
+  [adapter _conn]
+  (debug "Closing Discord" (a/uuid adapter))
+  (stop-connection! (:message @_conn))
+  (close!           (:event @_conn)))
+
+(defrecord Discord
+           [config
+            bot-id
+            conn
+            connected?
+            connection-last-active-timestamp
+            connection-latency
+            should-ping?
+            yetibot-user]
+
+  a/Adapter
+
+  (a/uuid [_] (:name config))
+
+  (a/platform-name [_] "Discord")
+
+  (a/channels [a] (channels a))
+
+  (a/send-paste [a msg] (send-msg a msg conn))
+
+  (a/send-msg [a msg] (send-msg a msg conn))
+
+  (a/join [_ channel]
+    (str
+     "Discord bots such as myself can't join channels on their own. Use "
+     "/invite from the channel you'd like me to join instead.✌️"))
+
+  (a/leave [_ channel]
+    (str
+     "Discord bots such as myself can't leave channels on their own. Use "
+     "/kick from the channel you'd like me to leave instead. 👊"))
+
+  (a/chat-source [_ channel] (chat-source channel))
+
+  (a/stop [adapter] (stop adapter conn))
+
+  (a/connected? [{:keys [connected?]}]
+    @connected?)
+
+  (a/connection-last-active-timestamp [_]
+    @connection-last-active-timestamp)
+
+  (a/connection-latency [_]
+    @connection-latency)
+
+  (a/start [adapter]
+    (start adapter conn config connected? bot-id yetibot-user)))
```

Comment:
Actually wasn't able to do this since the `conn, config, etc.` were atoms and I ended up getting an error around trying to reset an atom that is nil. 

I was trying to do this with:

```
{:keys [conn config connected? bot-id yetibot-user] :as adapter}
```

I could have done something wrong though or maybe there's another way to do this that I'm not aware of.

Reply:
Doing this seems to have worked

```
  [{conn :conn
    config :config
    connected? :connected?
    bot-id :bot-id
    yetibot-user :yetibot-user :as adapter}]
```

Maybe I just can't use `:keys` with atoms perhaps?